src/lib/core/data-table/data-table.ts

```diff
@@ -0,0 +1,323 @@
+import {
+  ChangeDetectionStrategy,
+  ChangeDetectorRef,
+  Component,
+  ContentChild,
+  ContentChildren,
+  Directive,
+  ElementRef,
+  Input,
+  IterableChangeRecord,
+  IterableChanges,
+  IterableDiffer,
+  IterableDiffers,
+  QueryList,
+  Renderer,
+  TemplateRef,
+  ViewChild,
+  ViewContainerRef,
+  ViewEncapsulation
+} from '@angular/core';
+import {BehaviorSubject} from 'rxjs/BehaviorSubject';
+import 'rxjs/add/operator/let';
+import 'rxjs/add/operator/debounceTime';
+import 'rxjs/add/observable/combineLatest';
+import {DataSource} from './data-source';
+
+export interface CollectionViewer {
+  start: number;
+  end: number;
+}
+
+/**
+ * Row cell definition for a CDK data-table.
+ * Captures the template of a column's data row cell as well as cell-specific properties.
+ */
+@Directive({selector: '[cdkRowCellDef]'})
+export class CdkRowCellDef {
+  constructor(public template: TemplateRef<any>) { }
+}
+
+
+/**
+ * Header row cell definition for a CDK data-table.
+ * Captures the template of a column's header cell and as well as cell-specific properties.
+ */
+@Directive({selector: '[cdkHeaderCellDef]'})
+export class CdkHeaderCellDef {
+  constructor(public template: TemplateRef<any>) { }
+}
+
+/**
+ * Column definition for the CDK data-table.
+ * Captures the template for the header and data cells of a column.
+ */
+@Directive({selector: '[cdkColumnDef]'})
+export class CdkColumnDef {
+  @Input('cdkColumnDef') name: string;
+
+  @ContentChild(CdkRowCellDef) cell: CdkRowCellDef;
+  @ContentChild(CdkHeaderCellDef) headerCell: CdkHeaderCellDef;
+}
+
+/**
+ * Header row definition for the CDK data-table.
+ * Captures the header row's template and other header properties such as the columns to display.
+ */
+@Directive({selector: '[cdkHeaderRowDef]'})
+export class CdkHeaderRowDef {
+  @Input('cdkHeaderRowDef') columns: string[];
+
+  constructor(public template: TemplateRef<any>) { }
+}
+
+/**
+ * Data row definition for the CDK data-table.
+ * Captures the header row's template and other row properties such as the columns to display.
+ */
+@Directive({selector: '[cdkRowDef]'})
+export class CdkRowDef {
+  @Input('cdkRowDefColumns') columns: string[];
+
+  // TODO: Add an input for providing a switch function to determine
+  // if this template should be used.
+
+  constructor(public template: TemplateRef<any>) { }
+}
+
+/** Header template container that contains the cell outlet. Adds the right class and role. */
+@Component({
+  selector: 'cdk-header-row',
+  template: '<ng-container cdkCellOutlet></ng-container>',
+  host: {
+    'class': 'cdk-header-row',
+    'role': 'row',
+  },
+  changeDetection: ChangeDetectionStrategy.OnPush,
+})
+export class CdkHeaderRow { }
```

Comment:
Might as well move each of these directives to their own file; the table itself will eventually get more complicated and it will be less churn to move them now